crates/sui-types/src/passkey_authenticator.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) 2021, Facebook, Inc. and its affiliates
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+use crate::crypto::PublicKey;
+use crate::signature_verification::VerifiedDigestCache;
+use crate::{
+    base_types::{EpochId, SuiAddress},
+    crypto::{DefaultHash, Signature, SignatureScheme, SuiSignature},
+    digests::ZKLoginInputsDigest,
+    error::{SuiError, SuiResult},
+    signature::{AuthenticatorTrait, VerifyParams},
+};
+use fastcrypto::hash::{HashFunction, Sha256};
+use fastcrypto::rsa::{Base64UrlUnpadded, Encoding};
+use fastcrypto::secp256r1::{Secp256r1PublicKey, Secp256r1Signature};
+use fastcrypto::traits::VerifyingKey;
+use fastcrypto::{error::FastCryptoError, traits::ToFromBytes};
+use once_cell::sync::OnceCell;
+use passkey::types::webauthn::{ClientDataType, CollectedClientData};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use shared_crypto::intent::IntentMessage;
+use std::hash::Hash;
+use std::hash::Hasher;
+use std::sync::Arc;
+
+#[cfg(test)]
+#[path = "unit_tests/passkey_authenticator_test.rs"]
+mod passkey_authenticator_test;
+
+/// An passkey authenticator with all the necessary fields.
+#[derive(Debug, Clone, JsonSchema, Serialize, Deserialize)]
+pub struct PasskeyAuthenticator {
+    authenticator_data: Vec<u8>,
+    client_data_json: Vec<u8>,
+    user_signature: Signature,
+    #[serde(skip)]
+    parsed_challenge: Vec<u8>,
+    #[serde(skip)]
+    bytes: OnceCell<Vec<u8>>,
+}
+
+impl PasskeyAuthenticator {
+    /// Validate client_data_json and initialize parsed challenge.
+    pub fn init(&mut self) -> Result<Self, SuiError> {
```

Comment:
The ability to have a PasskeyAuthenticator that never had `init()` called makes me quite nervous. I think it would be best to do these checks in a deserialize impl.

As an alternative to writing your own deserialize impl, you can also use the `#[serde(try_from = "T")]` feature, which would look something like this:

```
#[derive(Debug, Clone, JsonSchema, Serialize, Deserialize)]
#[serde(try_from = "PasskeyAuthenticatorParsed")]
#[serde(into = "PasskeyAuthenticatorParsed")]
pub struct PasskeyAuthenticator {
   ...
}

struct PasskeyAuthenticatorParsed {
     authenticator_data: Vec<u8>,
     client_data_json: Vec<u8>,
     user_signature: Signature,
}

impl TryFrom<PasskeyAuthenticatorParsed> for PasskeyAuthenticator {
   // implement try_from and put all checks here
}

impl From<PasskeyAuthenticator> for PasskeyAuthenticatorParsed {
   // convert back to unverified form for serialization
}
```